includes/data-port/class-sensei-data-port-utilities.php

```diff
@@ -26,12 +26,130 @@ public static function create_user( $username, $email = '' ) {
+	 *
+	 * @param string $source   Filename or URL.
+	 * @param int    $post_id  Id of the post.
+	 *
+	 * @return bool|WP_Error
+	 */
+	public static function attach_image_to_post( $source, $post_id ) {
+		if ( false === filter_var( $source, FILTER_VALIDATE_URL ) ) {
+
+			$attachments = get_posts(
+				[
+					'fields'         => 'ids',
+					'post_type'      => 'attachment',
+					'posts_per_page' => 1,
+					'post_status'    => 'any',
+					'meta_compare'   => 'REGEXP',
+					'meta_key'       => '_wp_attached_file', // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_key -- No faster way to search an attachment from its filename.
+					'meta_value'     => sanitize_file_name( $source ) . '$', // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_value -- See above.
+				]
+			);
+
+			if ( empty( $attachments ) ) {
+				return new WP_Error(
+					'sensei_data_port_attachment_not_found',
+					__( 'No attachment with the specified file name was found.', 'sensei-lms' )
+				);
+			}
+
+			$attachment_id = $attachments[0];
+		} else {
+			// In case a local URL is provided, try to convert it to the attachment.
+			$attachment_id = attachment_url_to_postid( $source );
+
+			if ( ! $attachment_id ) {
+				$attachment_id = self::create_attachment_from_url( $source );
+			}
+		}
+
+		if ( is_wp_error( $attachment_id ) ) {
+			return $attachment_id;
+		}
+
+		update_post_meta( $post_id, '_thumbnail_id', $attachment_id );
+
+		return true;
+	}
+
+	/**
+	 * This method retrieves a file from an external url, creates an attachment and links the attachment with the
+	 * downloaded file. If the file has been already downloaded an linked to an attachment, it returns the existing
+	 * attachment instead.
+	 *
+	 * @param string $external_url  The external url.
+	 *
+	 * @return int|WP_Error  The attachment id or an error.
+	 */
+	public static function create_attachment_from_url( $external_url ) {
+
+		$existing_attachment = get_posts(
+			[
+				'fields'         => 'ids',
+				'post_type'      => 'attachment',
+				'posts_per_page' => 1,
+				'title'          => md5( $external_url ),
+				'post_status'    => 'private',
+			]
+		);
+
+		if ( ! empty( $existing_attachment ) ) {
+			return $existing_attachment[0];
+		}
+
+		$response = wp_safe_remote_get( $external_url );
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$upload_result = wp_upload_bits( basename( $external_url ), null, wp_remote_retrieve_body( $response ) );
+
+		if ( ! empty( $upload_result['error'] ) ) {
+			return new WP_Error( 'sensei_data_port_storing_file_failure', $upload_result['error'] );
+		}
+
+		$file_path = $upload_result['file'];
+		$file_url  = $upload_result['url'];
+
+		$wp_filetype = wp_check_filetype_and_ext( $file_path, basename( $file_path ) );
+
+		$attachment_args = [
+			'post_content'   => basename( $file_path ),
+			'post_title'     => md5( $external_url ),
```

Comment:
I tried to avoid a slow query in the meta table but the truth is that there is an unavoidable one in `attach_image_to_post`. ea659bd6f972808f8c9f8128cf98c051aa8b1483